doc/guides/porting/arch.rst

```diff
@@ -478,38 +568,189 @@ To enable the :option:`CONFIG_HW_STACK_PROTECTION` feature, the system must
 
-If the target platform enables paging and requires drivers to memory-map
-their I/O regions, :option:`CONFIG_MMU` needs to be enabled and the
-:cpp:func:`arch_mem_map()` API implemented.
+Enabling user mode activates two new requirements:
+
+* A separate fixed-sized privilege mode stack, specified by
+  :option:`CONFIG_PRIVILEGED_STACK_SIZE`, must be allocated that the user
+  thread cannot access. It is used as the stack by the kernel when handling
+  system calls. If stack guards are implemented, a stack guard region must
+  be able to be placed before it, with support for carve-outs if necessary.
+
+* The memory protection hardware must be able to program a region that exactly
+  covers the thread's stack buffer, tracked in ``thread.stack_info``. This
+  implies that :c:macro:`ARCH_THREAD_STACK_SIZE_ADJUST()` will need to round
+  up the requested stack size so that a region may cover it, and that
+  :c:macro:`ARCH_THREAD_STACK_OBJ_ALIGN()` is also specified per the
+  granularity of the memory protection hardware.
+
+This become more complicated if the memory protection hardware requires that
+all memory regions be sized to a power of two, and aligned to their own size.
+This is common on older MPUs and is known with
+:option:`CONFIG_MPU_REQUIRES_POWER_OF_TWO_ALIGNMENT`.
+
+``thread.stack_info`` always tracks the user-accessible part of the stack
+object, it must always be correct to program a memory protection region with
+user access using the range stored within.
+
+Non power-of-two memory region requirements
+-------------------------------------------
+
+On systems without power-of-two region requirements, the reserved memory area
+for threads stacks defined by :c:macro:`K_THREAD_STACK_RESERVED` may be used to
+contain the privilege mode stack. The layout could be something like:
+
+::
+
+   +------------------------------+ <- thread.stack_obj
+   | Other platform data          |
+   +------------------------------+
+   | Guard region (if enabled)    |
+   +------------------------------+
+   | Guard carve-out (if needed)  |
+   |..............................|
+   | Privilege elevation stack    |
+   +------------------------------| <- thread.stack_obj +
+   | Stack buffer                 |      K_THREAD_STACK_RESERVED =
+   .                              .      thread.stack_info.start
+
+The guard region, and any carve-out (if needed) would be configured as a
+read-only region when the thread is created.
+
+* If the thread is a supervisor thread, the privilege elevation region is just
+  extra stack memory. An overflow will eventually crash into the guard region.
+
+* If the thread is running in user mode, a memory protection region will be
+  configured to allow user threads access to the stack buffer, but nothing
+  before or after it. An overflow in user mode will crash into the privilege
+  elevation stack, which the user thread has no access to. An overflow when
+  handling a system call will crash into the guard region.
```

Comment:
Maybe worth stating that in _this_ case we not only detect the overflows, but protect the kernel from corruption. Maybe implicit, or maybe worth mentioning

Reply:
I don't think this deep into an example of one way to configure the stack object layout is the best place to expand on this topic.

Reply:
I can put some detail in the toplevel "HW-based stack overflow detection" section though

Reply:
I added "Stack overflows in user mode are recoverable (from the kernel's perspective)
and require no special configuration; :option:`CONFIG_HW_STACK_PROTECTION`
only applies to catching overflows when the CPU is in sueprvisor mode.